conwayste/src/ui/pane.rs

```diff
@@ -16,56 +16,313 @@
  *  along with conwayste.  If not, see
  *  <http://www.gnu.org/licenses/>. */
 
+use std::error::Error;
 use std::fmt;
 
-use ggez::graphics::{self, Color, Rect, DrawMode, DrawParam};
+use ggez::graphics::{self, Color, DrawMode, DrawParam, Rect};
 use ggez::nalgebra::{Point2, Vector2};
 use ggez::{Context, GameResult};
 
+use enum_iterator::IntoEnumIterator;
 use id_tree::NodeId;
 
 use super::{
+    common::within_widget,
+    context,
+    focus::{CycleType, FocusCycle},
     widget::Widget,
-    common::{within_widget},
     UIError, UIResult,
 };
 
+use ggez::input::keyboard::KeyCode;
+
+use context::{EmitEvent, Event, EventType, Handled, UIContext};
+
 use crate::constants::colors::*;
 
 pub struct Pane {
     id: Option<NodeId>,
     z_index: usize,
     pub dimensions: Rect,
-    pub hover: bool,
     pub floating: bool, // can the window be dragged around?
     pub previous_pos: Option<Point2<f32>>,
     pub border: f32,
     pub bg_color: Option<Color>,
+    pub focus_cycle: FocusCycle,
+    pub handler_data: context::HandlerData, // required for impl_emit_event!
 
-    // might need something to track mouse state to see if we are still clicked within the
-    // boundaries of the pane in the dragging case
+                                            // might need something to track mouse state to see if
+                                            // we are still clicked within the boundaries of the
+                                            // pane in the dragging case
 }
 
 impl fmt::Debug for Pane {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "Pane {{ id: {:?}, z-index: {}, Dimensions: {:?} }}", self.id, self.z_index, self.dimensions)
+        write!(
+            f,
+            "Pane {{ id: {:?}, z-index: {}, Dimensions: {:?} }}",
+            self.id, self.z_index, self.dimensions
+        )
     }
 }
 
 /// A container of one or more widgets
 impl Pane {
     /// Specify the dimensional bounds of the Pane container
     pub fn new(dimensions: Rect) -> Self {
-        Pane {
+        let mut pane = Pane {
             id: None,
             z_index: std::usize::MAX,
-            dimensions: dimensions,
-            hover: false,
+            dimensions,
             floating: true,
             previous_pos: None,
             border: 1.0,
             bg_color: None,
+            focus_cycle: FocusCycle::new(CycleType::OpenEnded),
+            handler_data: context::HandlerData::new(),
+        };
+
+        // for each event type, define a handler of the appropriate type (mouse or keyboard)
+        for event_type in EventType::into_enum_iter() {
+            if event_type.is_mouse_event() {
+                let handler = |_obj: &mut dyn EmitEvent,
+                               uictx: &mut context::UIContext,
+                               evt: &context::Event|
+                 -> Result<Handled, Box<dyn Error>> {
```

Comment:
Same here regarding the handler in the body versus the `key_press_handler` down below.